src/sentry/utils/kafka.py

```diff
@@ -1,11 +1,43 @@
 import logging
 import signal
+import time
+from datetime import datetime
+from typing import Optional
+
+from sentry import options
 
 logger = logging.getLogger(__name__)
 
 
-def run_processor_with_signals(processor):
+def delay_kafka_rebalance(configured_delay: int) -> None:
+    """
+    Introduces a configurable delay to the consumer topic
+    subscription and consumer shutdown steps (handled by the
+    StreamProcessor). The idea behind is that by forcing
+    these steps to occur at certain time "ticks" (for example, at
+    every 15 second tick in a minute), we can reduce the number of
+    rebalances that are triggered during a deploy. This means
+    fewer "stop the world rebalancing" occurrences and more time
+    for the consumer group to stabilize and make progress.
+    """
+    now = float(datetime.now().strftime("%S.%f"))
+
+    next_tick, remainder = divmod(now, configured_delay)
+    if remainder > 0:
+        next_tick += 1
+
+    seconds_sleep = (configured_delay * next_tick) - now
+
+    time.sleep(seconds_sleep)
+
+
+def run_processor_with_signals(processor, consumer_name: Optional[str] = None):
     def handler(signum, frame):
+        if consumer_name == "ingest-generic-metrics" and options.get(
```

Comment:
I think you should be able to run this stuff in the `__init__` function of the indexer's strategy factory (not in `create_with_partitions`). then you don't have to add code here

Reply:
actually, in `__init__` and `shutdown` of the strategy factory

Reply:
Are you referring to the init and shutdown in the StreamProcessor? I'm not sure which shutdown you are referring to in the strategy factory?

Reply:
yep, ftr i realized `strategyfactory.shutdown` runs _after_ `consumer.close` in arroyo, and it would have to run before for that to make sense

Reply:
i just moved the `delay` call to the `init` of the metrics strategy factory (didn't make it configurable for all strategy factories). does this line up with what you were thinking as well?

Reply:
I just realized, I don't think this approach will work. We want to delay the initialization of the `StreamProcessor` because the subscribe step happens in the `__init__` (https://github.com/getsentry/arroyo/blob/main/arroyo/processing/processor.py#L283-L285). I think the delay needs to happen in the runner before we `get_stream_processor`, right?

Reply:
Never mind, I also just realized that the strategy factory gets initialized _before_ the StreamProcessor does. So your point (wrt moving the delay to the strategy factory) would still work.